src/devices/bus/a2bus/a2wico_trackball.cpp

```diff
@@ -0,0 +1,248 @@
+
+/*-------------------------------------------------
+    input_ports - device-specific input ports
+-------------------------------------------------*/
+
+ioport_constructor a2bus_wicotrackball_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(wicotrackball);
+}
+
+/*-------------------------------------------------
+    device_add_mconfig - device-specific
+    machine configurations
+-------------------------------------------------*/
+
+void a2bus_wicotrackball_device::device_add_mconfig(machine_config &config)
+{
+}
+
+/***************************************************************************
+    LIVE DEVICE
+***************************************************************************/
+
+a2bus_wicotrackball_device::a2bus_wicotrackball_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, type, tag, owner, clock),
+	device_a2bus_card_interface(mconfig, *this),
+	m_wicotrackballb(*this, WICOTRACKBALL_BUTTONS_TAG), m_wicotrackballxy(*this, { WICOTRACKBALL_XAXIS_TAG, WICOTRACKBALL_YAXIS_TAG }),
+	m_speed{ false, false }, m_buttons{ 0 }, m_wraparound{false}, m_axis{ 0, 0 },
+	m_last_pos{ WICOTRACKBALL_POS_UNINIT, WICOTRACKBALL_POS_UNINIT }
+{
+}
+
+a2bus_wicotrackball_device::a2bus_wicotrackball_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	a2bus_wicotrackball_device(mconfig, A2BUS_WICOTRACKBALL, tag, owner, clock)
+{
+}
+
+/*-------------------------------------------------
+    device_start - device-specific startup
+-------------------------------------------------*/
+
+void a2bus_wicotrackball_device::device_start()
+{
+	// register save state variables
+	save_item(NAME(m_speed));
+	save_item(NAME(m_wraparound));
+	save_item(NAME(m_buttons));
+	save_item(NAME(m_axis));
+	save_item(NAME(m_last_pos));
+}
+
+void a2bus_wicotrackball_device::device_reset()
+{
+	m_speed[0] = m_speed[1] = false;
+	m_wraparound = false;
+	m_buttons = 0;
+	m_axis[0] = m_axis[1] = 0;
+}
+
+uint8_t a2bus_wicotrackball_device::read_position(int axis)
+{
+	int speed_scale = (1 << m_speed[0]) << (m_speed[1] * 2);
+	uint8_t temp_axis = 0;
+	int diff_pos = 0;
+        int cur_pos = m_wicotrackballxy[axis]->read();
+
+	if (m_last_pos[axis] == WICOTRACKBALL_POS_UNINIT) {
+		m_last_pos[axis] = cur_pos;
+	}
+	diff_pos = cur_pos - m_last_pos[axis];
+	// wrap-around the positoin
+	if (diff_pos > 0x7f) {
+		diff_pos -= 0x100;
+	} else if (diff_pos < -0x80) {
+		diff_pos += 0x100;
+	}
+
+	if (m_wraparound) {
+		temp_axis = m_axis[axis] + diff_pos / speed_scale;;
+	} else {
+		if ((m_axis[axis] + diff_pos / speed_scale) < 0) {
+			temp_axis = 0;
+		} else if ((m_axis[axis] + diff_pos / speed_scale) > 0xff) {
+			temp_axis = 0xff;
+		} else {
+			temp_axis = m_axis[axis] + diff_pos / speed_scale;
+		}
+	}
+	m_last_pos[axis] = cur_pos;
+	m_axis[axis] = temp_axis;
+
+	return m_axis[axis];
+}
```

Comment:
This is called from a read handler and has side effects, which means it will wreak havoc when debugger memory windows are open.